pipeline/data_store.py

```diff
@@ -474,34 +491,34 @@ def get_provenance(self, process, pars_dict, upstream_provs=None, session=None):
             # check if we can find the upstream provenances
             upstreams = []
             for name in UPSTREAM_STEPS[process]:
+                prov = None
                 # first try to load an upstream that was given explicitly:
-                obj_names = PROCESS_PRODUCTS[name]
-                if isinstance(obj_names, str):
-                    obj_names = [obj_names]
-                obj = getattr(self, obj_names[0], None)  # only need one object to get the provenance
-                if isinstance(obj, list):
-                    obj = obj[0]  # for cutouts or measurements just use the first one
-                if upstream_provs is not None and name in [p.process for p in upstream_provs]:
-                    prov = [p for p in upstream_provs if p.process == name][0]
-
-                # second, try to get a provenance from objects saved to the store:
-                elif obj is not None and hasattr(obj, 'provenance') and obj.provenance is not None:
-                    prov = obj.provenance
-
-                # last, try to get the latest provenance from the database:
-                else:
-                    prov = get_latest_provenance(name, session=session)
+                if self.prov_tree is not None and name in self.prov_tree:
+                    prov = self.prov_tree[name]
+
+                if prov is None:  # if that fails, see if the correct object exists in memory
+                    obj_names = PROCESS_PRODUCTS[name]
+                    if isinstance(obj_names, str):
+                        obj_names = [obj_names]
+                    obj = getattr(self, obj_names[0], None)  # only need one object to get the provenance
+                    if isinstance(obj, list):
+                        obj = obj[0]  # for cutouts or measurements just use the first one
 
-                # can't find any provenance upstream, therefore
-                # there can't be any provenance for this process
-                if prov is None:
-                    return None
+                    if obj is not None and hasattr(obj, 'provenance') and obj.provenance is not None:
+                        prov = obj.provenance
 
-                upstreams.append(prov)
+                if prov is not None:  # if we don't find one of the upstreams, it will raise an exception
+                    upstreams.append(prov)
 
             if len(upstreams) != len(UPSTREAM_STEPS[process]):
                 raise ValueError(f'Could not find all upstream provenances for process {process}.')
 
+            for u in upstreams:  # check if "reference" is in the list, if so, replace it with its upstreams
```

Comment:
I need the subtraction image's provenance to have access to the provenance of the reference image and its products, so that it can load them when needed from the DB. It is the same as having the provenance of the new image and its products. 
If I just leave the upstream as the provenance of the reference it doesn't help in that regards.